### PR TITLE
Expose fetch() and XHR response bodies to the driver for access.

### PR DESCRIPTION
**Pull request by loganfsmyth**
This still uses the "sequence" concept I have in the backend PR, but since most of the actual important logic here isn't really reliant on a specific driver API, I figured this is fine to post as-is to get a head-start on review.

Refs https://github.com/RecordReplay/backend/issues/3676

### CODE REVIEW

File: dom/fetch/FetchDriver.cpp [hunk 1]
```diff
@@ -808,9 +810,9 @@ nsresult FetchDriver::HttpFetch(
       MOZ_ASSERT(!mAltDataListener);
       mAltDataListener = new AlternativeDataStreamListener(
           this, chan, aPreferredAlternativeDataType);
-      rv = chan->AsyncOpen(mAltDataListener);
+      chanListener = mAltDataListener;
```
Comment (loganfsmyth) [inline dom/fetch/FetchDriver.cpp:813]: I couldn't easily tell what "alternate data" in this context even was so I just opted to not record it. We can always add it if we find a case where it matters though.

File: mfbt/RecordReplay.h [hunk 2]
```diff
@@ -9,6 +9,7 @@
 #ifndef mozilla_RecordReplay_h
 #define mozilla_RecordReplay_h
 
+#include "mozilla/AlreadyAddRefed.h"
```
Comment (loganfsmyth) [inline mfbt/RecordReplay.h:12]: Hopefully adding this is fine, it's a pretty small header.

File: toolkit/recordreplay/JSControl.cpp [hunk 3]
```diff
@@ -616,7 +622,7 @@ static bool Method_OnHttpRequest(JSContext* aCx, unsigned aArgc, Value* aVp) {
   nsAutoCString requestId;
   ConvertJSStringToCString(aCx, args.get(0).toString(), requestId);
 
-  double bookmark = args.get(1).toDouble();
+  double bookmark = args.get(1).toNumber();
```
Comment (loganfsmyth) [inline toolkit/recordreplay/JSControl.cpp:625]: Turns out this was wrong and I was just lucky before!

File: mfbt/RecordReplay.h [hunk 4]
```diff
@@ -350,6 +352,7 @@ const char* GetBuildId();
 void OnTestCommand(const char* aString);
 void OnRepaintNeeded(const char* aWhy);
 bool IsTearingDownProcess();
+already_AddRefed<nsIStreamListener> WrapNetworkStreamListener(nsIStreamListener* listener);
```
Comment (bhackett1024) [inline mfbt/RecordReplay.h:358]: I know the lack of comments on the methods above this is pretty lame, but if you can add one on this method that would be great.

File: toolkit/recordreplay/JSControl.cpp (hunk 3 above)
Comment (bhackett1024) [reply toolkit/recordreplay/JSControl.cpp:625]: Yeah, toNumber() is best to use in spidermonkey to handle both the int32 and double cases.

File: toolkit/recordreplay/Network.cpp [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+#include "nsIPipe.h"
+#include "nsIChannel.h"
+#include "nsIInputStreamTee.h"
+#include "nsIStreamListenerTee.h"
+
+#include "mozilla/ToString.h"
+
+namespace mozilla::recordreplay {
+
+class ResponseRequestObserver final : public nsIRequestObserver {
```
Comment (bhackett1024) [inline toolkit/recordreplay/Network.cpp:31]: A comment above this class would be nice to have.

**Review by bhackett1024 — APPROVED** [review]

**Pull request merged**